rfcs/text/0003_application_service_mounting.md

```diff
@@ -0,0 +1,259 @@
+- Start Date: 2019-05-10
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+A front-end service to manage registration and root-level routing for
+first-class applications.
+
+# Basic example
+
+```tsx
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import { MyApp } from './components';
+
+class MyPlugin {
+  setup({ application }) {
+    application.registerApp({
+      id: 'my-app',
+      title: 'My Application',
+      rootRoute: '/myapp',
+      mount(targetDomElement, pluginStartContext, pluginStart) {
```

Comment:
What are `pluginStartContext` and `pluginStart`? Are those just `core` and `deps` that we get from `.setup` method? Basically, is that equivalent to:

```js
const MyPlugin = {
  setup: (core, deps) => {
    core.registerApp({
      mount: () => ReactDOM.render(<MyPlugin core={core} deps={deps} />
    });
  },
};
```

Reply:
Yes, these types are what the Migration guide refers to as `core: CoreSetup, plugins: PluginsSetup`, we have an outstanding chore to fix this discrepancy.

Reply:
OK, I think I understood this part, so the extra `core` and `plugins` objects provided to the `mount(el, core, plugins)` function is because they are different than the ones provided to the setup function `setup(core, plugins)` itself.

These `core` and `plugins` in `mount(el, core, plugins)` are the *start* life-cycle objects. But you need to inject them here in `mount()` function because otherwise the app cannot work with just the `core` from *setup* `setup(core, plugins)` life-cycle, because `core` across life-cycles are different.

This feels a bit confusing.

1. Maybe `core` could always be the same across the life-cycles? And if some method is not supposed to be called in some life-cycle you would raise an exception.

```js
// Same core across life-cycles:
class MyPlugin {
  setup () {
    // this.core
    // this.plugins
  }

  start () {
    // this.core
    // this.plugins
  }
}
```

2. Maybe app registration could be done in *start* life-cycle at any time? Then there is no need to inject `core` into the `mount()` method as the *start* life-cycle `core` would be already available.

```js
start (core) {
  core.application.register({
    // ...
    mount: el => render(<MyApp core={core} />, el),
  });
}
```

3. Maybe app registration and app mounting could be split between life-cycles?

```js
const plugin = {
  setup: setupCore => {
    setupCore.application.register({
      id: 'my-app',
      title: 'My Application',
      rootRoute: '/myapp',
    });
  },

  start: startCore => {
    startCore.application.mount('my-app', el => render(<MyApp core={startCore} />, el));
  },
};
```

Reply:
I'm opting to solve this problem with the Handler Context pattern instead. Option (3) here still could be useful if we think leaking `CoreSetup` into the mount function is going to cause us problems.